File: apps/android/feature/sets/domain/src/main/java/com/scribblefit/feature/sets/domain/usecase/AddSetToExerciseUseCase.kt
package com.scribblefit.feature.sets.domain.usecase

import com.scribblefit.core.common.runCatchingWithCancellation
import com.scribblefit.core.model.Exercise
import com.scribblefit.core.model.Set
import com.scribblefit.feature.sets.domain.SetRepository
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.withContext

/**
 * Appends a new set to an exercise.
 */
class AddSetToExerciseUseCase(
    private val repository: SetRepository,
    private val coroutineDispatcher: CoroutineDispatcher,
) {
    suspend operator fun invoke(exercise: Exercise): Result<Long> =
        withContext(coroutineDispatcher) {
            runCatchingWithCancellation {
                val exerciseId = exercise.id
                val nextSetNumber = (exercise.sets.maxOfOrNull { it.setNumber } ?: 0) + 1
                val newSet = Set(
                    id = 0,
                    setNumber = nextSetNumber,
                    reps = 0,
                    weight = 0f
                )
                repository.addSet(exerciseId, newSet)
            }
        }
}

File: apps/ios/ScribbleFit/LocalPackages/Features/FeatureExercises/Sources/FeatureExercises/Domain/UseCases/AddSetToExerciseUseCase.swift
import Foundation
import CoreModel

/**
 * Appends a new set to an exercise.
 */
public struct AddSetToExerciseUseCase: Sendable {
    private let exerciseRepository: ExerciseRepository

    public init(exerciseRepository: ExerciseRepository) {
        self.exerciseRepository = exerciseRepository
    }

    @MainActor
    public func execute(exerciseId: UUID, nextSetNumber: Int) async throws {
        guard let exercise = try await exerciseRepository.getExercise(id: exerciseId) else {
            return
        }

        let newSet = ExerciseSet(
            id: UUID(),
            setNumber: nextSetNumber,
            weight: 0.0,
            reps: 0
        )

        var updatedSets = exercise.sets
        updatedSets.append(newSet)

        try await exerciseRepository.updateExercise(exercise.copy(sets: updatedSets))
    }
}
